controllers/storagecluster/cephnfs.go

```diff
@@ -0,0 +1,340 @@
+}
+
+type ocsCephNFSBlockPool struct{}
+
+// newCephNFSBlockPoolInstance returns the cephBlockPool instance required for cephNFS that should be created
+// on first run.
+func (r *StorageClusterReconciler) newCephNFSBlockPoolInstance(initData *ocsv1.StorageCluster) (*cephv1.CephBlockPool, error) {
+	obj := &cephv1.CephBlockPool{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      generateNameForCephNFSBlockPool(initData),
+			Namespace: initData.Namespace,
+		},
+		Spec: cephv1.NamedBlockPoolSpec{
+			Name: ".nfs",
+			PoolSpec: cephv1.PoolSpec{
+				FailureDomain:  getFailureDomain(initData),
+				Replicated:     generateCephReplicatedSpec(initData, "data"),
+				EnableRBDStats: true,
+			},
+		},
+	}
+
+	err := controllerutil.SetControllerReference(initData, obj, r.Scheme)
+	if err != nil {
+		r.Log.Error(err, "Unable to set controller reference for CephBlockPool.", "CephBlockPool", klog.KRef(obj.Namespace, obj.Name))
+		return nil, err
+	}
+
+	return obj, nil
+}
+
+// ensureCreated ensures that CephNFS related CephBlockPool resource exist in the desired state.
+func (obj *ocsCephNFSBlockPool) ensureCreated(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) (reconcile.Result, error) {
+	if instance.Spec.NFS == nil || !instance.Spec.NFS.Enable {
+		return reconcile.Result{}, nil
+	}
+
+	cephBlockPool, err := r.newCephNFSBlockPoolInstance(instance)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	existingBlockPool := cephv1.CephBlockPool{}
+	err = r.Client.Get(context.TODO(), types.NamespacedName{Name: cephBlockPool.Name, Namespace: cephBlockPool.Namespace}, &existingBlockPool)
+
+	switch {
+	case err == nil:
+		if existingBlockPool.DeletionTimestamp != nil {
+			r.Log.Info("Unable to restore CephBlockPool because it is marked for deletion.", "CephBlockPool", klog.KRef(existingBlockPool.Namespace, existingBlockPool.Name))
+			return reconcile.Result{}, fmt.Errorf("failed to restore initialization object %s because it is marked for deletion", existingBlockPool.Name)
+		}
+
+		r.Log.Info("Restoring original CephBlockPool.", "CephBlockPool", klog.KRef(cephBlockPool.Namespace, cephBlockPool.Name))
+		existingBlockPool.ObjectMeta.OwnerReferences = cephBlockPool.ObjectMeta.OwnerReferences
+		existingBlockPool.Spec = cephBlockPool.Spec
+		err = r.Client.Update(context.TODO(), &existingBlockPool)
+		if err != nil {
+			r.Log.Error(err, "Failed to update CephBlockPool.", "CephBlockPool", klog.KRef(cephBlockPool.Namespace, cephBlockPool.Name))
+			return reconcile.Result{}, err
+		}
+	case errors.IsNotFound(err):
+		r.Log.Info("Creating CephBlockPool.", "CephBlockPool", klog.KRef(cephBlockPool.Namespace, cephBlockPool.Name))
+		err = r.Client.Create(context.TODO(), cephBlockPool)
+		if err != nil {
+			r.Log.Error(err, "Failed to create CephBlockPool.", "CephBlockPool", klog.KRef(cephBlockPool.Namespace, cephBlockPool.Name))
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{}, nil
```

Comment:
in case of other errors we are returning nil?

Reply:
> in case of other errors we are returning nil?

That's the pattern being following in other types too 
https://github.com/red-hat-storage/ocs-operator/blob/49c7a58a2d1ae83caa456366db6bab5691132ec0/controllers/storagecluster/cephblockpools.go#L126

https://github.com/red-hat-storage/ocs-operator/blob/49c7a58a2d1ae83caa456366db6bab5691132ec0/controllers/storagecluster/cephfilesystem.go#L106

Only error nil and notfound cases are handled.

Reply:
@jarrpa is this "pattern" intentional, or is it a mistake? 

If it's a mistake, we should not repeat it again here.

Reply:
Added default case to handle error now.